tests/utils/features/state.py

```diff
@@ -0,0 +1,157 @@
+        except KeyError:
+            raise KeyError('change {} not found in state.json'.format(id))
+
+    def get_task(self, id: str) -> dict[str, Any]:
+        try:
+            return self.state[TASKS][id]
+        except KeyError:
+            raise KeyError('task {} not found in state.json'.format(id))
+
+    def get_snap_type(self, snap_name: str) -> str:
+        '''
+        Given a snap name, returns the type of snap. If the snap name is not present,
+        returns the string "NOT_FOUND: " followed by the snap name
+        '''
+        if _keys_in_dict(self.state, DATA, SNAPS, snap_name, TYPE):
+            return self.state[DATA][SNAPS][snap_name][TYPE]
+        for task in self.state[TASKS].values():
+            if _keys_in_dict(task, DATA, SNAP_SETUP, SIDE_INFO, NAME) \
+                    and task[DATA][SNAP_SETUP][SIDE_INFO][NAME] == snap_name:
+                return task[DATA][SNAP_SETUP][TYPE]
+
+        return "NOT_FOUND: {}".format(snap_name)
+
+    def get_snap_types_from_task_id(self, id: str) -> set[str]:
+        '''
+        Retrieves the type of snap associated with the task with the given id.
+        If the task kind is present in the exception list or if the task does
+        not have a data section, then returns an empty set.
+
+        :raises KeyError: when the snap type was not found yet the kind of task is not in the exception list
+        '''
+
+        task = self.get_task(id)
+        if task['kind'] in TASKS_WITHOUT_SNAP:
+            return {}
+        if 'data' not in task:
+            return {}
+        data = task['data']
+
+        if _keys_in_dict(data, SNAP_TYPE):
+            return {data[SNAP_TYPE]}
+
+        elif _keys_in_dict(data, SNAP_SETUP, TYPE):
+            return {data[SNAP_SETUP][TYPE]}
+
+        elif _keys_in_dict(data, SNAP_SETUP, SIDE_INFO, NAME):
+            return {self.get_snap_type(data[SNAP_SETUP][SIDE_INFO][NAME])}
+
+        elif _keys_in_dict(data, SNAP_SETUP_TASK):
+            return self.get_snap_types_from_task_id(data[SNAP_SETUP_TASK])
+
+        elif _keys_in_dict(data, HOOK_SETUP, SNAP):
+            return {self.get_snap_type(data[HOOK_SETUP][SNAP])}
+
+        elif _keys_in_dict(data, PLUG, SNAP) and _keys_in_dict(data, SLOT, SNAP):
+            return {self.get_snap_type(data[PLUG][SNAP]), self.get_snap_type(data[SLOT][SNAP])}
+
+        elif _keys_in_dict(data, SNAPS):  # ex: conditional-auto-refresh
+            return {snap_data[TYPE] for snap_data in data[SNAPS].values()}
+
+        elif _keys_in_dict(data, SERVICE_ACTION, SNAP_NAME):
+            return {self.get_snap_type(data[SERVICE_ACTION][SNAP_NAME])}
+
+        elif _keys_in_dict(data, QUOTA_CONTROL_ACTIONS):
+            if any(action for action in data[QUOTA_CONTROL_ACTIONS] if SNAPS in action):
+                return {self.get_snap_type(snap) for action in data[QUOTA_CONTROL_ACTIONS] if SNAPS in action for snap in action[SNAPS]}
+            return {}
```

Comment:
I'm would leave the types this kind of tasks for the time being, right now we expect them to be mostly app/gadget and then it is immaterial. You can leave a comment about them though